src/utils.ml

```diff
@@ -152,7 +154,11 @@ module Ast_io = struct
   let read input_source ~input_kind =
     try
       match input_source with
-      | Stdin -> from_channel stdin ~input_kind
+      | Stdin ->
+          (match input_kind with
+          | Necessarily_binary -> set_binary_mode_in stdin true
+          | _ -> ());
+          from_channel stdin ~input_kind
```

Comment:
The default text mode translates `\r\n` to `\n`, so a magic number like `Caml1999M033` should have no problem being read in text mode. In general, as long as there is no byte `0x0d` (CR) in the magic number it should be safe to read in text mode, and then switch to binary mode.

The only thing I am worried about is buffering. Would the entire buffer be translated (big problem), rather than just the magic number field (no problem). Perhaps @nojb can comment?

Reply:
> The only thing I am worried about is buffering. Would the entire buffer be translated (big problem), rather than just the magic number field (no problem). Perhaps @nojb can comment?

As mentioned above, by the time the bytes get into the input buffer, translation has already ocurred, so it is too late to switch modes after checking the magic number. Switching `stdin` to binary mode before reading the magic seems the only way. And anyway, binary mode works just fine for reading source code.